kotlin/workflow-ui-core/src/main/java/com/squareup/workflow/ui/Compatible.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.workflow.ui
+
+/**
+ * Returns true if [me] and [you] are instances of the same class, unless
+ * that class implements [Compatible]. In that case
+ * `me.`[canUpdateFrom][Compatible.isCompatibleWith]`(you)` must also be true.
+ *
+ * A convenient way to take control over the matching behavior of objects that
+ * don't implement [Compatible] is to wrap them with [Uniqued].
+ */
+fun compatible(
+  me: Any,
+  you: Any
+): Boolean {
+  return when {
+    me::class != you::class -> false
+    me !is Compatible<*> -> true
+    // If you see a casting error here, it's a lie. https://youtrack.jetbrains.com/issue/KT-31823
+    else -> me.isCompatibleWith(you as Compatible<*>)
```

Comment:
Could you avoid the cast (and address my above comment) by doing:
```kotlin
when {
  me is Compatible<*> && you is Compatible<*> -> me.isCompatibleWith(you)
  else -> me::class == you::class
}
```

Reply:
This isn't the contract. They must be of the same type _and_ return true from this method. Remember the recursive type param on `isCompatibleWith()`, we promise only to make the call if both left and right are instances of `T`.